### ISSUE TITLE
Android Auto Support

### ISSUE THREAD

**Issue by azothForgotten** [izderadicka/audioserve-android#5]
I started working on making the app work with android auto. For the most part there is a simple configuration change that makes it available as an app, but when it's opened on Android Auto there are no MediaItems found. Currently playing items will show up and you can navigate the current queue, but you can't navigate to any books. I think this has to do with the way the folders are built out. I'm not super familiar with Android development so it's a bit new to me. Do you have any plans to support Android Auto?

**Comment by izderadicka** [comment]
No plans for Android Auto in near future.
This was my first android app, so it may happen, that not all things are perfect.
MediaSession is a complex beast, I used Exoplayer's implementation as start, but I guess there still might be some glitches.  
Initially I looked at https://github.com/googlesamples/android-UniversalMusicPlayer  as inspiration, which should have all these Android thingies (like Auto) right, but might get bit distracted during the implementation.

You may start looking at permissions to use media browser across apps, as I think I initially simplified it to be used just for this app, broader usage requires some more settings, authentication .

**Comment by izderadicka** [comment]
I tested quickly on latest emulator (with Android Q), which has Android Auto 4.1 built in.  But I'm not sure what should happen there -  if audioserve is playing in background I can't see anything in Android Auto.

> Currently playing items will show up and you can navigate the current queue

I do not see anything like this - looks like audioserve client is completely ignored by Android Auto.
So what actually is expected to happen there?

I did not study much the integration with AA, basic official documentation is here - https://developer.android.com/training/cars/media,   looking at it quickly:
* Android Auto support declaration has to be in app manifest, which is not now
* Android Auto is using MediaBrowserCompat and MediaSessionCompat, we are using that too, BUT,
* As both are pretty complex with tons of options and posibilities not exactly sure if we are using it in exactly right way in all cases (for sure search is not compatible) - to work with Android Auto some refactoring and rework of our implementation (which is in AudioService.kt file) will be needed.

As stated this is not priority for me, but will see in future, maybe somebody else would like to look at it?

**Comment by azothForgotten** [comment]
Yeah to get it show up in AA I had to modify the manifest and set the exported value to true on the audio service.
![image](https://user-images.githubusercontent.com/13576951/59218951-3c904380-8b76-11e9-9ebb-1d6ea16889f4.png)

If I get more time I will tinker with it and hopefully send a pull request at some point.

Thanks for responding so quickly!

**Comment by azothForgotten** [comment]
Made a pull request that seems to fix everything for me. I really only had to change the onGetRoot function in the AudioService. I left it mostly the way you had it for simplicity's sake so at the moment the if statement is a bit pointless. I wasn't really sure of the original purpose, so I left it. #6

**Comment by izderadicka** [comment]
The original purpose was to provide some kind on authorization, who can access media browser, which somehow got lost as service was not exported.   This is a "proper" way from above mentioned Universal Music Player:

```kotlin

/**
     * Returns the "root" media ID that the client should request to get the list of
     * [MediaItem]s to browse/play.
     */
    override fun onGetRoot(
        clientPackageName: String,
        clientUid: Int,
        rootHints: Bundle?
    ): BrowserRoot? {

        /*
         * By default, all known clients are permitted to search, but only tell unknown callers
         * about search if permitted by the [BrowseTree].
         */
        val isKnownCaller = packageValidator.isKnownCaller(clientPackageName, clientUid)
        val rootExtras = Bundle().apply {
            putBoolean(
                MEDIA_SEARCH_SUPPORTED,
                isKnownCaller || browseTree.searchableByUnknownCaller
            )
            putBoolean(CONTENT_STYLE_SUPPORTED, true)
            putInt(CONTENT_STYLE_BROWSABLE_HINT, CONTENT_STYLE_GRID)
            putInt(CONTENT_STYLE_PLAYABLE_HINT, CONTENT_STYLE_LIST)
        }

        return if (isKnownCaller) {
            // The caller is allowed to browse, so return the root.
            BrowserRoot(UAMP_BROWSABLE_ROOT, rootExtras)
        } else {
            /**
             * Unknown caller. There are two main ways to handle this:
             * 1) Return a root without any content, which still allows the connecting client
             * to issue commands.
             * 2) Return `null`, which will cause the system to disconnect the app.
             *
             * UAMP takes the first approach for a variety of reasons, but both are valid
             * options.
             */
            BrowserRoot(UAMP_EMPTY_ROOT, rootExtras)
        }
}
```

packageValidator then uses this list of known applications https://github.com/googlesamples/android-UniversalMusicPlayer/blob/master/common/src/main/res/xml/allowed_media_browser_callers.xml

I think we'll need to work on it a bit in a way similar to above, not sure if we want every application to have access
